core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParser.scala

```diff
@@ -25,7 +27,9 @@ trait SlickExpressionParser extends JavaTokenParsers {
   // entity following references. The final step takes the last entity, producing a result dependent on the query
   // (e.g. loading the entity itself, getting an attribute). The root step produces an entity to start the pipeline.
   // If rootStep is None, steps should also be empty and finalStep does all the work.
-  case class PipelineQuery(rootStep: Option[RootFunc], steps: List[PipeFunc], finalStep: FinalFunc)
+  abstract class PipelineQuery[FR <: FinalResult[_]](val rootStep: Option[RootFunc], val steps: List[PipeFunc], val finalStep: FinalFunc[FR])
+  case class EntityPipelineQuery(override val rootStep: Option[RootFunc], override val steps: List[PipeFunc], override val finalStep: EntityFinalFunc) extends PipelineQuery(rootStep, steps, finalStep)
+  case class AttributePipelineQuery(override val rootStep: Option[RootFunc], override val steps: List[PipeFunc], override val finalStep: AttributeFinalFunc) extends PipelineQuery(rootStep, steps, finalStep)
```

Comment:
I've split out the query pipeline into separate entity and attribute versions. They were being called separately, going through the same parser, and then forking again at the other end.